### PR TITLE
Fix type definitions left over from Jasmine 1.x

### PR DESCRIPTION
**Pull request by sgravrock**
This removes type definitions for things that were either removed or made unavailable in Jasmine 2.0, and fixes a couple of related mismatches between the type definitions and Jasmine itself. 

Jasmine 1.x didn't have a well-defined public API, and as a result much of the type definition file describes what are now private APIs. I've left those alone for now as long as (a) they're correct, and (b) it's plausible that a user could access them and do something useful with them. So e.g. I've left in the parts of Env that are now private but removed most of Spec (fails (a) and all of TreeBuilder (fails (b)).

- [*] Use a meaningful title for the pull request. Include the name of the package modified.
- [*] Test the change in your own code. (Compile and run.)
- [*] [Add or edit tests](https://github.com/DefinitelyTyped/DefinitelyTyped/blob/master/README.md#my-package-teststs) to reflect the change.
- [*] Follow the advice from the [readme](https://github.com/DefinitelyTyped/DefinitelyTyped/blob/master/README.md#make-a-pull-request).
- [*] Avoid [common mistakes](https://github.com/DefinitelyTyped/DefinitelyTyped/blob/master/README.md#common-mistakes).
- [*] [Run `npm test <package to test>`](https://github.com/DefinitelyTyped/DefinitelyTyped/blob/master/README.md#running-tests).

Select one of these and delete the others:
If changing an existing definition:
- [*] Provide a URL to documentation or source code which provides context for the suggested changes: https://jasmine.github.io/api/3.7/global
- [ ] If this PR brings the type definitions up to date with a new version of the JS library, update the version number in the header.

### CODE REVIEW

**Comment by chivesrs** [comment]
There's a lot going on in this PR. I'm happy to review it, but please @ me only after you get the tests to pass both locally and in the GitHub CI environment. It looks like by removing stuff in Jasmine, you are causing other packages to break.

Maybe consider putting this PR back into draft mode, fixing all the other packages first to use the non deprecated equivalents of whatever, then come back to this?

**Comment by chivesrs** [comment]
I also noticed you're a core maintainer of Jasmine, so I trust that all the changes you make are legit, I won't bother looking into that.

One way of proceeding is, you delete the fields that you are able to delete that don't break the tests, and the ones that you want to delete but are unable to, due to other packages relying on them, mark them with the `@deprecated` JSDoc comment. Then you can submit a follow up PR that cleans up those little by little.

**Comment by sgravrock** [comment]
I'll dig through the build failures and assess whether they represent actual breakage. My expectation is that it's impossible for user code to both depend on the definitions I'm removing and also work against Jasmine >= 2.0, but I'll want to assess each failure. This might take a while, though, because it looks like I'm only getting one failure per CI build. You advised running all the tests locally. That seems like the thing to do, but I'm not sure how to do it. I tried running `npm run test-all` locally but I found that it sometimes passed, sometimes fails with the following (even on master with no changes) and never produces the errors I get from the CI build.

```
$ npm run test-all

> definitely-typed@0.0.3 test-all /Users/steve/src/DefinitelyTyped
> node node_modules/@definitelytyped/dtslint-runner/dist/index.js --path .

Using local DefinitelyTyped at .
Parsing definitions...
Found 7579 packages.
Parsing in 8 child process(es)...
Parsing took 25.125 s
Error: Unexpected key in not-needed package: typingsPackageName
    at new NotNeededPackage (/Users/steve/src/DefinitelyTyped/node_modules/@definitelytyped/definitions-parser/dist/packages.js:183:23)
    at /Users/steve/src/DefinitelyTyped/node_modules/@definitelytyped/definitions-parser/dist/packages.js:354:66
    at Array.map (<anonymous>)
    at Object.readNotNeededPackages (/Users/steve/src/DefinitelyTyped/node_modules/@definitelytyped/definitions-parser/dist/packages.js:354:45)
    at Object.parseDefinitions (/Users/steve/src/DefinitelyTyped/node_modules/@definitelytyped/definitions-parser/dist/parse-definitions.js:39:60)
    at async Object.prepareAffectedPackages (/Users/steve/src/DefinitelyTyped/node_modules/@definitelytyped/dtslint-runner/dist/prepareAffectedPackages.js:16:5)
    at async Object.runDTSLint (/Users/steve/src/DefinitelyTyped/node_modules/@definitelytyped/dtslint-runner/dist/main.js:32:11)
    at async /Users/steve/src/DefinitelyTyped/node_modules/@definitelytyped/dtslint-runner/dist/index.js:97:26
```

I'm using Node 14.12.0, in case it matters.

**Comment by chivesrs** [comment]
I've never tried to run all tests (or run all affected tests), let me copy your PR and try.

**Comment by chivesrs** [comment]
Just as an FYI, the CI seems to be using

```
Run actions/setup-node@v1
/opt/hostedtoolcache/node/12.21.0/x64/bin/node --version
v12.21.0
/opt/hostedtoolcache/node/12.21.0/x64/bin/npm --version
6.14.11
```

and my machine is using

```
chives@chives-VirtualBox:~/jasmine$ node --version
v12.21.0
chives@chives-VirtualBox:~/jasmine$ npm --version
6.14.11
```

I highly doubt the node version matters, but just pasting this for reference.

**Comment by chivesrs** [comment]
This error seems to be coming due to the fact your fork+branch are not rebased off the current HEAD of DT
(error being `Error: Unexpected key in not-needed package: typingsPackageName`)

Once I did that I ran into a new error `Error: RangeError [ERR_CHILD_PROCESS_STDIO_MAXBUFFER]: stdout maxBuffer length exceeded`

Instead I'm going to change ideas and try to run only affected tests. The CI output gave me this nice line:
```
Testing 2 changed packages: jasmine,jasminewd2
Testing 21 dependent packages: angular-wizard,frisby v0.8,gulp-jasmine,jasmine_dom_matchers,jasmine-ajax,jasmine-data_driven_tests,jasmine-data-provider,jasmine-enzyme,jasmine-es6-promise-matchers,jasmine-expect-jsx,jasmine-fixture,jasmine-jquery,jasmine-matchers,jasmine-node,jasmine-promise-matchers,jasminewd2,karma-jasmine,node-expat,protractor-helpers,saywhen,testing-library__jasmine-dom
```

```
chives@chives-VirtualBox:~/jasmine/DefinitelyTyped$ IN="angular-wizard,frisby v0.8,gulp-jasmine,jasmine_dom_matchers,jasmine-ajax,jasmine-data_driven_tests,jasmine-data-provider,jasmine-enzyme,jasmine-es6-promise-matchers,jasmine-expect-jsx,jasmine-fixture,jasmine-jquery,jasmine-matchers,jasmine-node,jasmine-promise-matchers,jasminewd2,karma-jasmine,node-expat,protractor-helpers,saywhen,testing-library__jasmine-dom"
chives@chives-VirtualBox:~/jasmine/DefinitelyTyped$ IFS=","
chives@chives-VirtualBox:~/jasmine/DefinitelyTyped$ for x in $IN; do npm test $x; done
```

Some packages lint cleanly, some show the error (I can see the this one `Namespace 'jasmine' has no exported member 'Runner'.` in my output), and some fail to lint because I haven't done some npm install commands. There was one that due to having a space it completely bugged out.

**Comment by chivesrs** [comment]
Regarding frisby I think `chives@chives-VirtualBox:~/jasmine/DefinitelyTyped$ npm test frisby/v0` seems to do the trick.

It looks like jasmine-node is really your only issue, and the rest of the packages are fine. So the dependency hell looks to be very small and manageable.

**Comment by chivesrs** [comment]
> because it looks like I'm only getting one failure per CI build

I think the reason for that is because you in fact only had one failing package. And presumably also only one failing line, usually I see the CI show you all failing lines, but not 100% on that second statement.

**Comment by chivesrs** [comment]
Regarding the tests passing on master with 0 changes, I think the way DT has set things up is that only the most recent commit's affected packages are tested, not every single package in the repo. I'm not 100% sure on this though, so asking a DT maintainer is your best bet. Nonetheless, I think this is a moot point now, now that we figured out how to read the CI output to find all affected packages, and any packages you directly edited.

**Comment by UziTech** [comment]
For any packages that should reference older versions of jasmine you can change them to reference [/types/jasmine/v1/](https://github.com/DefinitelyTyped/DefinitelyTyped/blob/master/types/jasmine/v1/) or [/types/jasmine/v2/](https://github.com/DefinitelyTyped/DefinitelyTyped/blob/master/types/jasmine/v2/) as described in https://github.com/DefinitelyTyped/DefinitelyTyped#if-a-library-is-updated-to-a-new-major-version-with-breaking-changes-how-should-i-update-its-type-declaration-package

**Comment by sgravrock** [comment]
@UziTech Thanks for the pointer. I'm having some trouble updating jasmine-node after changing it to refer to jasmine/v1. It appears that switching from a triple-slash reference to an import (in db4bc0f562) has some consequences:

```
$ npm run test jasmine-node

> definitely-typed@0.0.3 test /Users/steve/src/DefinitelyTyped
> dtslint types "jasmine-node"

Error: /Users/steve/src/DefinitelyTyped/types/jasmine-node/index.d.ts:9:1
ERROR: 9:1    strict-export-declare-modifiers  'declare' keyword is redundant here. See: https://github.com/Microsoft/dtslint/blob/master/docs/strict-export-declare-modifiers.md
ERROR: 9:18   strict-export-declare-modifiers  All declarations in this module are exported automatically. Prefer to explicitly write 'export' for clarity. If you have a good reason not to export this declaration, add 'export {}' to the module to shut off automatic exporting. See: https://github.com/Microsoft/dtslint/blob/master/docs/strict-export-declare-modifiers.md
ERROR: 11:1   strict-export-declare-modifiers  'declare' keyword is redundant here. See: https://github.com/Microsoft/dtslint/blob/master/docs/strict-export-declare-modifiers.md
ERROR: 11:19  strict-export-declare-modifiers  All declarations in this module are exported automatically. Prefer to explicitly write 'export' for clarity. If you have a good reason not to export this declaration, add 'export {}' to the module to shut off automatic exporting. See: https://github.com/Microsoft/dtslint/blob/master/docs/strict-export-declare-modifiers.md
ERROR: 18:39  expect                           TypeScript@4.3 compile error: 
Namespace '"/Users/steve/src/DefinitelyTyped/types/jasmine-node/index".jasmine' has no exported member 'Runner'.
ERROR: 42:5   expect                           TypeScript@4.3 compile error: 
Exports and export assignments are not permitted in module augmentations.

/Users/steve/src/DefinitelyTyped/types/jasmine-node/jasmine-node-tests.ts:6:50
ERROR: 6:50   expect                           TypeScript@4.3 compile error: 
Expected 2 arguments, but got 3.
```

I tried to fix those errors in 45995c9963 but I'm still stuck with one. I get the same error if I apply my jasmine-node changes to master, so I think it's not directly due to the changes I made to the Jasmine type definitions.

```
Error: /Users/steve/src/DefinitelyTyped/types/jasmine-node/index.d.ts:20:39
ERROR: 20:39  expect  TypeScript@4.3 compile error: 
Namespace 'jasmine' has no exported member 'Runner'.
```

Any suggestions? I'd also appreciate some extra eyes on my jasmine-node changes, since I'm running up against the limits of my TypeScript knowlege here.

**Comment by chivesrs** [comment]
I'll copy your fork and take a look at it in a bit.

That being said jasmine-node is marked as being a deprecated project so I'm not very worried about making breaking changes to it.

**Comment by chivesrs** [comment]
Here's a working index.d.ts and jasmine-node-tests.ts. That should pass the tests.

```
// Type definitions for jasmine-node v1.14.5
// Project: https://github.com/mhevery/jasmine-node
// Definitions by: Sven Reglitzki <https://github.com/svi3c>
// Definitions: https://github.com/DefinitelyTyped/DefinitelyTyped
// TypeScript Version: 3.9

import "jasmine";

declare global {
    function it(expectation: string, assertion: (done: (err?: any) => void) => void, timeout?: number): void;

    namespace jasmine {
        interface Env {
            defaultTimeoutInterval: number;
        }

        interface ExecuteSpecsOptions {
            specFolders: string[];
            onComplete?: (runner: jasmine.Runner) => void;
            isVerbose?: boolean;
            showColors?: boolean;
            teamcity?: string | boolean;
            useRequireJs?: boolean;
            regExpSpec: RegExp;
            junitreport?: {
                report: boolean;
                savePath: string;
                useDotNotation: boolean;
                consolidate: boolean;
            };
            includeStackTrace?: boolean;
            growl?: boolean;
        }

        interface JasmineNode {
            executeSpecsInFolder(options: ExecuteSpecsOptions): void;
            loadHelpersInFolder(path: string, pattern: RegExp): void;
        }
    }
}

declare const jasmine: jasmine.JasmineNode;
export = jasmine;
```

```
// Type definitions for jasmine-node v1.14.5
// Project: https://github.com/mhevery/jasmine-node
// Definitions by: Sven Reglitzki <https://github.com/svi3c/>
// Definitions: https://github.com/DefinitelyTyped/DefinitelyTyped

it("should have a timeout parameter", () => {}, 1000);
it("should have an optional timeout parameter", () => {});

import jasmine = require("jasmine-node");

jasmine.loadHelpersInFolder("root", /\.helper\.ts/);

jasmine.executeSpecsInFolder({
    specFolders: [],
    onComplete: runner => {
        console.log(runner.results().failedCount);
    },
    isVerbose: true,
    showColors: true,
    teamcity: false,
    useRequireJs: false,
    regExpSpec: /\.spec\.ts/,
    junitreport: {
        report: false,
        savePath: "./reports/",
        useDotNotation: true,
        consolidate: true,
    },
    includeStackTrace: true,
    growl: false,
});
```

**Comment by chivesrs** [comment]
CI looks to pass. Ship it?

**Comment by sgravrock** [comment]
Looks good to me. Thanks, @chivesrs.

**Review by chivesrs — APPROVED** [review]

**Comment by andrewbranch** [comment]
This is a pretty scary amount of deletions from an extremely popular package; I would feel much better if I could get one more owner to review.  @svi3c @borisyankov @theodorejb @davidparsson @gmoothart @lukas-zech-software @Engineer2B @cyungmann @Roaders @devoto13 @FDIM @kolodny @stephenfarrar @djungowski @kirjs @ienzam @sjelin @gkalpak anyone willing to help out? 🙏

**Comment by chivesrs** [comment]
@andrewbranch The author of this PR is a core jasmine maintainer. I would say that they are the most qualified person here to write these sorts of PRs. I'll let you respond back with your comments but I won't say the magic words for the bot to merge just yet.

**Comment by sgravrock** [comment]
Thanks for the vote of confidence. That said, it *is* "a pretty scary amount of deletions from an extremely popular package", and I certainly don't mind waiting to get an extra pair of eyes on it.

File: types/jasmine-node/index.d.ts
Comment (peterblazejewicz) [inline types/jasmine-node/index.d.ts:5]:
```suggestion
// TypeScript Version: 3.9
```
why this change? What's the new feature requiring this? The package itself is maintenance mode:
https://github.com/mhevery/jasmine-node#project-status

File: types/jasminewd2/jasminewd2-tests.ts
Comment (peterblazejewicz) [inline types/jasminewd2/jasminewd2-tests.ts:146]: we should probably add `paths` resolution, so the tests passes with older version, instead of removing this test, what do you think?

File: types/jasmine-node/jasmine-node-tests.ts
Comment (peterblazejewicz) [inline types/jasmine-node/jasmine-node-tests.ts:4]: btw, this could be dropped as no-op code (so dead one really)

File: types/jasminewd2/jasminewd2-tests.ts
Comment (sgravrock) [reply types/jasminewd2/jasminewd2-tests.ts:146]: I gave that a quick try just now and it looks like the switch from triple-slash require to import will probably require a significant rewrite of the definition file similar to what happened to jasmine-node in this PR. I think it's less risky to remove that one line. It looks like jasminewd2 (not my package, btw) always required Jasmine 2, and `jasmine.getEnv().currentSpec` only worked in Jasmine 1.

Comment (peterblazejewicz) [reply types/jasminewd2/jasminewd2-tests.ts:146]: Agree,thx!

File: types/jasmine/index.d.ts
Comment (sgravrock) [inline types/jasmine/index.d.ts:327]: I just double-checked all the deletions and found that this function and `Suite#getFullName` still exist. They're private APIs but it's possible that user code could be using them, at least in the case of `formatErrorMessage`. (I'm not sure it's even possible to get a reference to a `Suite` from outside Jasmine. Should I put those two methods back?

File: types/jasmine-node/index.d.ts
Comment (chivesrs) [reply types/jasmine-node/index.d.ts:5]: No idea. It seemed to be required for tests to pass. Something about merging the two global jasmine namespaces?
Yes I agree the project is in maintenance mode which is why I'm ok with making this somewhat breaking change.

File: types/jasmine/index.d.ts
Comment (chivesrs) [reply types/jasmine/index.d.ts:327]: If they're private APIs and you haven't documented them in the official jasmine docs, I'm ok with deleting them.

Comment (UziTech) [reply types/jasmine/index.d.ts:327]: It is possible to get the suites and specs from `jasmine.env.topSuite()` and it's children.

Comment (peterblazejewicz) [reply types/jasmine/index.d.ts:327]: Why not just add `@deprecated`? You can also extract all those methods into `legacy.d.ts` and incorporate in the module export (so you can phase those out incrementally)

File: types/jasmine-node/index.d.ts
Comment (peterblazejewicz) [reply types/jasmine-node/index.d.ts:5]:
```diff
diff --git a/types/jasmine-node/index.d.ts b/types/jasmine-node/index.d.ts
index 8000a8a986..71889e71ca 100644
--- a/types/jasmine-node/index.d.ts
+++ b/types/jasmine-node/index.d.ts
@@ -2,8 +2,6 @@
 // Project: https://github.com/mhevery/jasmine-node
 // Definitions by: Sven Reglitzki <https://github.com/svi3c>
 // Definitions: https://github.com/DefinitelyTyped/DefinitelyTyped
-// TypeScript Version: 3.9
-
 import "jasmine";
 
 declare global {
@@ -16,7 +14,7 @@ declare global {
 
         interface ExecuteSpecsOptions {
             specFolders: string[];
-            onComplete?: (runner: jasmine.Runner) => void;
+            onComplete?: (runner: Runner) => void;
             isVerbose?: boolean;
             showColors?: boolean;
             teamcity?: string | boolean;
```
this will fix tests, `jasmine` v1 is not a proper module)

Comment (peterblazejewicz) [reply types/jasmine-node/index.d.ts:5]: (removing version will make it compatible and tested with all TS tested via DT CI)

**Review by chivesrs — APPROVED** [review]

**Review by peterblazejewicz — APPROVED** [review]

File: types/jasmine/index.d.ts
Comment (UziTech) [inline types/jasmine/index.d.ts:862]: I don't know the best practices but I don't think this should be deprecated unless it is going to change soon. When using jasmine it is not very useful to get the Suite but it is very useful when creating a reporter.

Comment (sgravrock) [reply types/jasmine/index.d.ts:862]: Can you point me at an example? The thing that Jasmine passes to reporters [isn't actually a suite](https://jasmine.github.io/api/edge/global.html#SuiteResult), but maybe I'm not thinking of the same thing you are.

Comment (UziTech) [reply types/jasmine/index.d.ts:862]: [atom-jasmine3-test-runner](https://github.com/UziTech/atom-jasmine3-test-runner) needs to use the suite to display the suite description. We get the suites and specs from `jasmine.getEnv().topSuite()`.

image

Comment (sgravrock) [reply types/jasmine/index.d.ts:862]: Ok, that helps. It looks like `atom-jasmine3-test-runner` is building up UI based on the structure of the suites and specs before the first spec starts. That can't be done using just the reporter interface. I'll remove the deprecation. Longer term, I think the right move is to add the parts of Suite that are useful for this sort of thing to the Jasmine API docs, and mark the other members deprecated here.

Comment (peterblazejewicz) [reply types/jasmine/index.d.ts]: >I don't know the best practices but I don't think this should be deprecated unless it is going to change soon. 

This obviously would show in IDEs like VSCode as deprecated, which could trick users. I've proposed to consider this, if you're to remove that from the type definition itself for methods, which are identified as wrong. The `deprecated` can be not the best choice for this reason, though. thx!

Comment (sgravrock) [reply types/jasmine/index.d.ts]: @UziTech I took a quick pass at the doc additions. Let me know if you think anything else should be added: https://jasmine.github.io/api/edge/Suite.html

**Review by chivesrs — APPROVED** [review]

**Review by UziTech — APPROVED** [review]

File: types/jasmine/index.d.ts
Comment (UziTech) [reply types/jasmine/index.d.ts]: The docs look good to me.

**Review by rbuckton — APPROVED** [review]

**Comment by chivesrs** [comment]
Ready to merge

(wonder if this still works as a reviewer, or if it has to be PR owner)

**Comment by chivesrs** [comment]
Looks like PR owner only I guess, or the bot didn't like my extra characters after `Ready to merge`.

**Comment by peterblazejewicz** [comment]
Yup, @author only

**Comment by chivesrs** [comment]
They must have changed it then, I recall it being any reviewer as well.

**Review by peterblazejewicz — APPROVED** [review]

**Comment by peterblazejewicz** [comment]
Maybe that's because of failing CI, can you folks amend last commit message and force push a change? (or you could create empty commit to just pull CI trigger, you know the drill I think)

**Comment by sgravrock** [comment]
I pushed a commit with an amended message but no code changes to trigger a fresh build. Sorry about the delay.

**Comment by chivesrs** [comment]
That was the wrong move. The CI wasn't failing, it was benchmarks failing, which are irrelevant.
You should have just merged.
Now we have to wait again for a DT maintainer, probably another week or two wait.

**Review by chivesrs — APPROVED** [review]

**Review by UziTech — APPROVED** [review]

**Review by peterblazejewicz — APPROVED** [review]

**Comment by chivesrs** [comment]
Thank you so much Steve for this PR, the TypeScript/DT team for dealing with the back and forth, and all the reviewers for the thorough review. Appreciate it.

**Pull request merged**
